File: test/com/barrybecker4/puzzle/maze1/model/DistributionMap.java
package com.barrybecker4.puzzle.maze1.model;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

/**
 * @author Barry Becker
 */
public class DistributionMap extends HashMap<Direction, List<Integer>> {

    DistributionMap() {
        this.put(Direction.FORWARD, Arrays.asList(0, 0, 0));
        this.put(Direction.LEFT, Arrays.asList(0, 0, 0));
        this.put(Direction.RIGHT, Arrays.asList(0, 0, 0));
    }

    DistributionMap(List<Integer> forwardDist, List<Integer> leftDist, List<Integer> rightDist) {
        this.put(Direction.FORWARD, forwardDist);
        this.put(Direction.LEFT, leftDist);
        this.put(Direction.RIGHT, rightDist);
    }

    void increment(List<Direction> directions) {
        increment(directions.get(0), 0);
        increment(directions.get(1), 1);
        increment(directions.get(2), 2);
    }

    void increment(Direction dir, int position) {
        get(dir).set(position, get(dir).get(position) + 1);
    }
}

File: scala-test/com/barrybecker4/puzzle/maze/model/DistributionMap.scala
package com.barrybecker4.puzzle.maze.model

import com.barrybecker4.puzzle.maze.model.Direction.{FORWARD, LEFT, RIGHT}

/**
  * Immutable counts of where FORWARD, LEFT, RIGHT appeared in each of the three shuffled positions.
  */
case class DistributionMap(
    forwardDist: List[Int] = List(0, 0, 0),
    leftDist: List[Int] = List(0, 0, 0),
    rightDist: List[Int] = List(0, 0, 0)
) {

  def increment(directions: List[Direction]): DistributionMap =
    incrAt(directions(0), 0).incrAt(directions(1), 1).incrAt(directions(2), 2)

  private def incrAt(dir: Direction, position: Int): DistributionMap = dir match {
    case FORWARD => copy(forwardDist = forwardDist.updated(position, forwardDist(position) + 1))
    case LEFT    => copy(leftDist = leftDist.updated(position, leftDist(position) + 1))
    case RIGHT   => copy(rightDist = rightDist.updated(position, rightDist(position) + 1))
  }
}
